src/sage/combinat/partition.py

```diff
@@ -8784,14 +8930,198 @@ def cardinality(self):
+            sage: Partitions_length_and_parts_restricted(9, 3, 5, 2, 9)
+            Partitions of 9 having length between 3 and 5 and whose parts are at least 2
+        """
+        if self._min_length == 1 and self._max_length == self._n:
+            length_str = ""
+        elif self._min_length == self._max_length:
+            length_str = f"having length {self._min_length}"
+        elif self._min_length == 1:
+            length_str = f"having length at most {self._max_length}"
+        elif self._max_length == self._n:
+            length_str = f"having length at least {self._min_length}"
+        else:
+            length_str = f"having length between {self._min_length} and {self._max_length}"
+
+        if self._min_part == 1 and self._max_part == self._n:
+            parts_str = ""
+        elif self._min_part == self._max_part:
+            parts_str = f"having parts equal to {self._min_part}"
+        elif self._min_part == 1:
+            parts_str = f"whose parts are at most {self._max_part}"
+        elif self._max_part == self._n:
+            parts_str = f"whose parts are at least {self._min_part}"
+        else:
+            parts_str = f"whose parts are between {self._min_part} and {self._max_part}"
+
+        if length_str:
+            if parts_str:
+                return f"Partitions of {self._n} " + length_str + " and " + parts_str
+            return f"Partitions of {self._n} " + length_str
+        if parts_str:
+            return f"Partitions of {self._n} " + parts_str
+        return f"Partitions of {self._n}"
+
+    def __contains__(self, x):
+        """
+        Check if ``x`` is contained in ``self``.
+
+        TESTS::
+
+            sage: P = Partitions(10, min_part=2, max_part=5, min_length=2, max_length=4)
+            sage: Partition([]) in P
+            False
+
+            sage: Partition([3]) in P
+            False
+
+            sage: Partition([5, 3, 2]) in P
+            True
+        """
+        try:
+            mu = Partition(x)
+        except ValueError:
+            return False
+        return (mu.size() == self._n
+                and (not mu
+                     or (min(mu) >= self._min_part
+                         and max(mu) <= self._max_part
+                         and self._min_length <= len(mu) <= self._max_length)))
+
+    def __iter__(self):
+        """
+        Iterator over the set of partitions in ``self``.
+
+        EXAMPLES::
+
+            sage: list(Partitions(9, min_part=2, max_part=4, min_length=3, max_length=4))
+            [[4, 3, 2], [3, 3, 3], [3, 2, 2, 2]]
+        """
+        yield from IntegerListsLex(self._n, max_slope=0,
+                                   min_part=self._min_part,
+                                   max_part=self._max_part,
+                                   min_length=self._min_length,
+                                   max_length=self._max_length,
+                                   element_constructor=Partition)
```

Comment:
```suggestion
        yield from IntegerListsLex(self._n, max_slope=0,
                                   min_part=self._min_part,
                                   max_part=self._max_part,
                                   min_length=self._min_length,
                                   max_length=self._max_length,
                                   element_constructor=lambda x: self.element_class(self, x))
```
Otherwise this is yielding elements in the wrong parent.

Reply:
good catch!